crypto/ess/ess_lib.c

```diff
@@ -289,73 +289,93 @@ static int ess_issuer_serial_cmp(const ESS_ISSUER_SERIAL *is, const X509 *cert)
-
-    /* Recompute SHA1 hash of certificate if necessary (side effect). */
-    if (!x509v3_cache_extensions(cert))
-        return -1;
+    if (cid == NULL && cid_v2 == NULL) {
+        ERR_raise(ERR_LIB_ESS, ERR_R_PASSED_INVALID_ARGUMENT);
+        return -2;
+    }
 
-    /* TODO(3.0): fetch sha1 algorithm from providers */
-    if (!X509_digest(cert, EVP_sha1(), cert_sha1, NULL))
-        return -1;
+    /* Look for cert with cid in the certs. */
+    for (i = 0; i < sk_X509_num(certs); ++i) {
+        cert = sk_X509_value(certs, i);
 
-    /* Look for cert in the cert_ids vector. */
-    for (i = 0; i < sk_ESS_CERT_ID_num(cert_ids); ++i) {
-        const ESS_CERT_ID *cid = sk_ESS_CERT_ID_value(cert_ids, i);
+        /* TODO(3.0): fetch sha algorithm from providers */
+        if (cid != NULL)
+            md = EVP_sha1();
+        else
+            md = cid_v2->hash_alg == NULL ? EVP_sha256() :
+                EVP_get_digestbyobj(cid_v2->hash_alg->algorithm);
+        if (md == NULL) {
+            ERR_raise(ERR_LIB_ESS, ESS_R_ESS_DIGEST_ALG_UNKNOWN);
+            return -2;
+        }
 
-        if (cid->hash->length == SHA_DIGEST_LENGTH
-            && memcmp(cid->hash->data, cert_sha1, SHA_DIGEST_LENGTH) == 0) {
-            const ESS_ISSUER_SERIAL *is = cid->issuer_serial;
+        cid_hash_len = cid != NULL ? cid->hash->length : cid_v2->hash->length;
+        if (!X509_digest(cert, md, cert_digest, &len)
+                || cid_hash_len != len) {
+            ERR_raise(ERR_LIB_ESS, ESS_R_ESS_CERT_DIGEST_ERROR);
+            return -2;
+        }
 
-            if (is == NULL || ess_issuer_serial_cmp(is, cert) == 0)
-                return i;
+        if (memcmp(cid != NULL ? cid->hash->data : cid_v2->hash->data,
+                   cert_digest, len) == 0) {
+            is = cid != NULL ? cid->issuer_serial : cid_v2->issuer_serial;
+            /* Well, it's not really required to matching the serial numbers. */
+            if (is == NULL || ess_issuer_serial_cmp(is, cert) == 0) {
+                if ((i == 0) == (index == 0))
+                    return i;
+                ERR_raise(ERR_LIB_ESS, ESS_R_ESS_CERT_ID_WRONG_ORDER);
+                return -2;
+            }
         }
     }
 
+    ERR_raise(ERR_LIB_ESS, ESS_R_ESS_CERT_ID_NOT_FOUND);
     return -1;
 }
 
-/* Returns < 0 if certificate is not found, certificate index otherwise. */
-int ossl_ess_find_cert_v2(const STACK_OF(ESS_CERT_ID_V2) *cert_ids,
-                          const X509 *cert)
+/*
+ * If ESSCertID and/or ESSCertIDv2 exist, which must be non-empty if given,
+ * check if their first ID entry matches the signer cert first in chain
+ * and each further ID entry matches any further cert in the chain.
+ */
+int ossl_ess_check_signing_certs(const ESS_SIGNING_CERT *ss,
+                                 const ESS_SIGNING_CERT_V2 *ssv2,
+                                 const STACK_OF(X509) *chain,
+                                 int require_signingCertificate)
```

Comment:
Please drop this parameter. Apparently it is always required. And the name is super ugly.

Reply:
Having a closer look at this and related functions, 
I found that it should be exported as
`OSSL_ESS_check_signing_certs()` along with `OSSL_ESS_signing_cert{,v2}_new_init()`.

This makes the ESS lib usable also for other applications than TSP and CAdES-BES.
For this general use it is good to keep the option because other uses may not strictly require the presence of the `signingCertificate{,v2}` attribute.
I had deliberately chosen the name `require_signingCertificate` for consistency with the naming in the RFCs, but changed it to `require_signing_cert` with is in line with the OpenSSL naming conventions.

Reply:
But in https://mailarchive.ietf.org/arch/browse/spasm/?gbt=1&index=JOHi3aOBMNTYgZkCITsUcoM-JVQ Russ said that there must be at least one cert id in the sequence if the SigningCertificate is there.
So no, I do not think there should be such parameter.

Reply:
One needs to look very close here because the thing is a litle confusing. 
Russ wrote:
> RFC 2634 says that the signer's certificate MUST be included.  The attribute should be omitted if the sequence is empty.

My respective question and his response were about the ESSCertID{,v2} list, which is inside a signingCertificate attribute.
The `require_signing_cert` governs whether of the signingCertificate attribute is required, which is at a different level.

The existence of at least one cert in the ESSCertID{,v2} list is instead checked unconditionally by
```
    if (n_v1 == 0 || n_v2 == 0) {
        ERR_raise(ERR_LIB_ESS, ESS_R_EMPTY_ESS_CERT_ID_LIST);
        return 0;
    }
```

Reply:
OK then